subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java

```diff
@@ -1039,11 +1040,18 @@ private boolean wasElementRemovedAfterRealized() {
         }
 
         protected RuntimeException domainObjectCreationException(Throwable cause) {
-            return new IllegalStateException(String.format("Could not create domain object '%s' (%s)", getName(), getType().getSimpleName()), cause);
+            return new DomainObjectCreationException(DefaultNamedDomainObjectCollection.this.getDisplayName(), getName(), getType(), cause);
         }
     }
 
     private static RuntimeException domainObjectRemovedException(String name, Class<?> type) {
         return new IllegalStateException(String.format("The domain object '%s' (%s) for this provider is no longer present in its container.", name, type.getSimpleName()));
     }
+
+    @Contextual
+    private static class DomainObjectCreationException extends IllegalStateException {
```

Comment:
Any reason we chose a custom exception instead of `InvalidUserCodeException`, which is already `@Contextual`?

Reply:
Easy, I forgot that existed